### PR TITLE
Initialize repo

### PR DESCRIPTION
**Pull request by robertsLando**
@mcollina @gnought Could this be the right way? 

I would like to create all tests using async/await, so I have created an helper with client/broker promisified functions (and some utils). 

About persistences deps, what about add them as git repo instead of using npm repo? 
This will allow us to make tests before publishing a persistence in npm.

Also tests will be done using github actions and env vars. The env vars will decide the persistence to use: 

```js
var persistence = require(process.env.AEDES_PERSISTENCE)
var mqemitter = require(process.env.MQEMITTER)
```

Are you ok with this approach?

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "aedes-tests",
+  "version": "0.0.1",
+  "description": "Integration/Black Box tests for Aedes MQTT Broker",
+  "main": "index.js",
+  "scripts": {
+    "unit": "tap --no-esm -J test/*.js",
```
Comment (mcollina) [inline package.json]: can you use a .taprc instead? I recently found out how useful they are.
Also disable flow, jsx and typescript.

File: package.json [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    "test": "npm run unit"
```
Comment (mcollina) [inline package.json]: can you please add standard as well?

File: package.json [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  },
+  "homepage": "https://github.com/moscajs/aedes-tests#readme",
+  "devDependencies": {
+    "aedes": "^0.40.1",
+    "aedes-persistence-mongodb": "^8.0.0",
+    "aedes-persistence-redis": "^7.0.0",
+    "mqtt": "^3.0.0",
+    "standard": "^14.3.3",
+    "tap": "^14.10.7",
+    "ws": "^7.2.3"
```
Comment (mcollina) [inline package.json:62]: we need the mqemitters as well.

File: test/helper.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
+const util = require('util')
```
Comment (mcollina) [inline test/helper.js]: use strict is missing

File: test/helper.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    })
+  } else {
+    server = net.createServer(options, aedesHandler)
+  }
+  return server
+}
+
+function subscribe (client, topic, options) {
+  return new Promise((resolve, reject) => {
+    client.subscribe(topic, options, function (err) {
```
Comment (mcollina) [inline test/helper.js]: please use util.promisify instead.

**Review by mcollina — CHANGES_REQUESTED** [review]

**Comment by mcollina** [comment]
I'm ok with using env vars. I don't think you can just require with them because you'll likely need to pass in some custom options for all of them.

So, I would use `DB=redis` or `DB=mongo` and fill it all from there instead of passing it to require.

File: package.json (hunk 1 above)
Comment (robertsLando) [reply package.json]: > Also disable flow, jsx and typescript.

Could you tell me how I can do this in package.json please?

**Comment by robertsLando** [comment]
> So, I would use DB=redis or DB=mongo and fill it all from there instead of passing it to require.

Ok!

And what about 

> About persistences deps, what about add them as git repo instead of using npm repo?
This will allow us to make tests before publishing a persistence in npm.

Example `npm install --save moscajs/aedes-persistence-redis#master`

**Comment by mcollina** [comment]
Let's cover published modules first, and then we see how to check in master directly.

**Comment by robertsLando** [comment]
Are you LGTM with this? @mcollina

**Review by mcollina — CHANGES_REQUESTED** [review]
I would prefer if we had a single test added

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,49 @@
+on: [push, pull_request]
+
+jobs:
+  test:
```
Comment (mcollina) [inline .github/workflows/ci.yml]: This should be called redis

File: test/basic.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
+'use-strict'
+
+const helper = require('./helper.js')
+const { test } = require('tap')
+
+test('Connect-Publish-Disconnect 1000 clients', async function (t) {
+  const aedes = helper.initAedes()
+  const server = await helper.createServer(aedes.handle)
```
Comment (mcollina) [inline test/basic.js]: We might also wrap those two into one.

File: test/basic.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  aedes.on('clientError', function (c, error) {
+    t.error('Client error')
+  })
+
+  aedes.on('publish', function (packet, client) {
+    // ignore $SYS topics
+    if (packet.topic === 'my/topic') {
+      published++
+    }
+  })
```
Comment (mcollina) [inline test/basic.js]: I would avoid to create aedes and test the public API. I would literally start it as a black box _in a separate process_.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: test/basic.js (hunk 8 above)
Comment (robertsLando) [reply test/basic.js]: Oh, ok. So should I create a file `broker.js` where I init aedes and with tls, ws and net support support and then create a test suites with clients that connect to that?

Comment (robertsLando) [reply test/basic.js]: I have made some changes. Let me know if that is the way, then I need to know how you would like to handle aedes process, should I kill it after all tests are ended?

File: test/basic.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
+'use-strict'
+
+const helper = require('./helper.js')
+const { test } = require('tap')
+
+const { fork } = require('child_process')
+
+const brokerProcess = fork('aedes.js')
```
Comment (mcollina) [inline test/basic.js]: I think you want to spin this up & down inside each test.

**Review by mcollina — CHANGES_REQUESTED** [review]
Good work, we are very close!

File: test/basic.js (hunk 9 above)
Comment (robertsLando) [reply test/basic.js]: Ok that is what I would like to know :)

**Comment by robertsLando** [comment]
@mcollina Now it should be almost ready! Should I remove logs from aedes process?

File: aedes.js [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+  cert: readFileSync('./server.cert'),
+  rejectUnauthorized: false
+}
+
+function listen (server, proto) {
+  return new Promise((resolve, reject) => {
+    server.listen(ports[proto], (err) => {
+      if (err) reject(err)
+      else {
+        console.log(proto, 'server is listening on port', ports[proto])
```
Comment (mcollina) [inline aedes.js]: I would remove all console.log. We might use aedes-logs if we have one, or rig something up with pino

File: aedes.js [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+// })
+
+// broker.on('unsubscribe', function (subscriptions, client) {
+//   console.log('MQTT client \x1b[32m' + (client ? client.id : client) +
+//         '\x1b[0m unsubscribed to topics: ' + subscriptions.join('\n'))
+// })
+
+// broker.on('publish', async function (packet, client) {
+//   console.log('Client \x1b[31m' + (client ? client.id : 'BROKER_' + broker.id) + '\x1b[0m has published', packet.payload.toString(), 'on', packet.topic)
+// })
```
Comment (mcollina) [inline aedes.js]: please remove these commented lines

File: test/helper.js [hunk 12]
```diff
@@ -0,0 +1,66 @@
+'use strict'
+
```
Comment (mcollina) [inline test/helper.js:2]: This file will be run by the wildcard in package.json. Can you rename `test/basic.js` into `test/basic.test.js` and adjust that pattern? Thanks

**Review by mcollina — CHANGES_REQUESTED** [review]

File: package.json [hunk 13]
```diff
@@ -0,0 +1,64 @@
+{
+  "name": "aedes-tests",
+  "version": "0.0.1",
+  "description": "Integration/Black Box tests for Aedes MQTT Broker",
+  "main": "index.js",
+  "scripts": {
+    "unit": "tap test/*.test.js",
```
Comment (mcollina) [inline package.json:7]: You'd need to disable parallel testing

**Review by mcollina — CHANGES_REQUESTED** [review]

File: package.json (hunk 13 above)
Comment (robertsLando) [reply package.json:7]: You right, done!

**Review by mcollina — APPROVED** [review]
lgtm

File: test/basic.test.js [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+  var publisher = await helper.startClient()
+
+  await publisher.publish(msg.topic, msg.payload, msg)
+  await helper.delay(2000)
+
+  await Promise.all(subscribers.map(c => c.end()))
+  await publisher.end()
+
+  t.tearDown(helper.closeBroker)
```
Comment (mcollina) [inline test/basic.test.js]: this line must be at the top of the test

**Review by mcollina — CHANGES_REQUESTED** [review]

**Comment by robertsLando** [comment]
@mcollina I wait for test to end then I merge. Could we create an issue with a list of the tests to implement?

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
> @mcollina I wait for test to end then I merge. Could we create an issue with a list of the tests to implement?

Sure thing, go ahead!

**Comment by robertsLando** [comment]
I have draft a list, if you have suggestions please just tell me: https://github.com/moscajs/aedes-tests/issues/2

**Pull request merged**